emukit/bayesian_optimization/acquisitions/constrained_bayesian_optimization.py

```diff
@@ -0,0 +1,68 @@
+from ...core.interfaces import IModel
+from ...core.loop import FixedIntervalUpdater, OuterLoop, SequentialPointCalculator
+from ...core.loop.loop_state import create_loop_state
+from ...core.optimization import AcquisitionOptimizer
+from ...core.parameter_space import ParameterSpace
+
+
+class ConstrainedBayesianOptimizationLoop(OuterLoop):
+
+    def __init__(self, space: ParameterSpace, 
+                 model_objective: IModel, 
+                 model_constraint: IModel,
+                 acquisition: Acquisition = None,
+                 update_interval: int = 1, 
+                 batch_size: int = 1):
+
+        """
+        Emukit class that implements a loop for building Bayesian optimization with an unknown constraint.
+        For more information see:
+
+        Michael A. Gelbart, Jasper Snoek, and Ryan P. Adams,
+        Bayesian Optimization with Unknown Constraints,
+        https://arxiv.org/pdf/1403.5607.pdf
+
+        :param space: Input space where the optimization is carried out.
+        :param model_objective: The model that approximates the underlying objective function
+        :param model_constraint: The model that approximates the unknown constraints
+        :param acquisition: The acquisition function that will be used to collect new points (default, CEI).
+        :param update_interval:  Number of iterations between optimization of model hyper-parameters. Defaults to 1.
+        """
+
+        if not np.all(np.isclose(model_objective.X, model_constraint.X)):
+            raise ValueError('Emukit currently only supports identical '
+                             'training inputs for the constrained and objective model')
+
+        if acquisition is None:
+            acquisition_objective = ExpectedImprovement(model_objective)
+            acquisition_constraint = ProbabilityOfFeasibility(model_constraint)
+            acquisition_constrained = acquisition_objective * acquisition_constraint
+
+
+        model_updater_objective = FixedIntervalUpdater(model_objective, update_interval)
+        model_updater_constraint = FixedIntervalUpdater(model_constraint, update_interval, lambda state: state.cost)
+
+        acquisition_optimizer = AcquisitionOptimizer(space)
+        if batch_size == 1:
+            candidate_point_calculator = SequentialPointCalculator(acquisition, acquisition_optimizer)
+        else:
+            if not isinstance(model, IDifferentiable):
+                raise ValueError('Model must implement ' + str(IDifferentiable) +
+                                 ' for use with Local Penalization batch method.')
+            log_acquisition = LogAcquisition(acquisition)
+            candidate_point_calculator = LocalPenalizationPointCalculator(log_acquisition, acquisition_optimizer, model,
+                                                                          space, batch_size)
+
+        loop_state = create_loop_state(model_objective.X, model_objective.Y, model_constraint.Y)
```

Comment:
So `model_constraint.Y` is treated as cost? I don't know if that's right or wrong, just clarifying for myself.

Reply:
I used cost because it was already available in loop_state,
otherwise I think I should have defined a new variable called e.g. y_constraint
Do you have any suggestion on this?

Reply:
I do have an idea, but that may be a bit more work. I think it is ok for loops to have their specific state classes. So here you would create ConstrainedLoopState, which would define additional fields you need.

In the hindsight this is probably how cost should have been implemented.

I'd also like to know what others think of this approach. @marpulli @mmahsereci thoughts?

Reply:
I may or may not have another example where a specific loop state would be quite handy @apaleyes. So I like that idea better. (generally speaking)